lib/internal/streams/readable.js

```diff
@@ -1284,9 +1322,265 @@ ObjectDefineProperties(ReadableState.prototype, {
     set(value) {
       this[kPaused] = !!value;
     }
+  },
+
+  // https://fetch.spec.whatwg.org/#dom-body-bodyused
+  bodyUsed: {
+    get() {
+      return isDisturbed(this);
+    }
+  },
+
+  body: {
+    get() {
+      if (this[kConsume]?.type === kWebStreamType) {
+        return this[kConsume].stream;
+      }
+
+      return consume(this, kWebStreamType);
+    }
+  },
+
+  text: {
+    get() {
+      return consume(this, kTextType);
+    }
+  },
+
+  json: {
+    get() {
+      return consume(this, kJSONType);
+    }
+  },
+
+  blob: {
+    get() {
+      return consume(this, kBlobType);
+    }
+  },
+
+  arrayBuffer: {
+    get() {
+      return consume(this, kArrayBufferType);
+    }
   }
 });
 
+function isLocked(self) {
+  return self[kConsume]?.stream?.locked === true;
+}
+
+// https://streams.spec.whatwg.org/#readablestream-disturbed
+function isDisturbed(self) {
+  return self.destroyed || self.readableDidRead;
+}
+
+// https://fetch.spec.whatwg.org/#body-unusable
+function isUnusable(self) {
+  return isDisturbed(self) || isLocked(self);
+}
+
+function consume(self, type) {
+  if (isUnusable(self)) {
+    // eslint-disable-next-line no-restricted-syntax
+    throw new TypeError('unusable');
+  }
+
+  if (type === kWebStreamType) {
+    if (!ReadableStream) {
+      ReadableStream = require('internal/webstreams/readablestream')
+        .ReadableStream;
+    }
+
+    const objectMode = self.readableObjectMode;
+    const highWaterMark = self.readableHighWaterMark;
+    // When not running in objectMode explicitly, we just fall
+    // back to a minimal strategy that just specifies the highWaterMark
+    // and no size algorithm. Using a ByteLengthQueuingStrategy here
+    // is unnecessary.
+    let strategy;
+    if (objectMode) {
+      if (!CountQueuingStrategy) {
+        CountQueuingStrategy = require('internal/webstreams/queuingstrategies');
+      }
+      strategy = new CountQueuingStrategy({ highWaterMark });
+    } else {
+      strategy = { highWaterMark };
+    }
+
+    self
+      .on('error', function(err) {
+        const { controller } = this[kConsume];
+        controller.error(err);
+      })
+      .on('close', function() {
+        const { controller } = this[kConsume];
+        if (controller) {
+          controller.error(new AbortError());
+        }
+      });
+    const consume = self[kConsume] = {
+      type,
+      objectMode,
+      controller: null,
+      push(chunk) {
+        const { objectMode, controller } = this;
+
+        assert(controller);
+
+        if (chunk === null) {
+          // TODO: Do we need to wait for construct to finish?
```

Comment:
Why would we need to wait for construct to finish?

Reply:
start can be async so I would assume the intent is that start should not return until the stream has started. I'm not quite sure what web streams have in mind in this regard...

Reply:
@jasnell maybe has some insight